Fix #92: rewrite samples using async await.

damondouglas opened the pull request.

damondouglas commented:
Hello @kwalrath please let me know if I could make this better and thank you for considering my contribution.

Sfshaza commented:
Hey, Damon, I'm the other Dart writer. THANKS for doing this! I'm sure we'll want an engineer to review, but we appreciate it!
--shams

kwalrath commented on dart_io_mini_samples/example/files_directories_and_symlinks/renaming_a_file_directory_or_symlink.dart:
```diff
@@ -8,18 +8,19 @@
 import 'dart:io';
 
-void main() {
+void main() async {
```
We should remove the "void", right?

kwalrath commented on dart_io_mini_samples/example/files_directories_and_symlinks/symlinks/checking_if_a_path_represents_a_symlink.dart:
```diff
@@ -9,10 +9,9 @@ void main() {
 
   // List the contents of the system temp directory.
   systemTempDir.list(recursive: true, followLinks: false)
```
Should this use await for instead?

kwalrath commented:
I'm only partway through looking at this, but thanks very much @damondouglas! I looked up your CLA status and you've signed it, so that's good.

damondouglas commented:
Ah I'm doing for Streams now.  Thank you.

damondouglas commented:
@kwalrath and @Sfshaza , Thank you again for considering my contribution.  I cleaned up a bit of void main's and implemented await for Streams.  I confess I didn't realize this was possible with Streams.  Dart is such a beautiful language.

kwalrath commented:
Great! I'll get back to reviewing this soon.

kwalrath commented:
OK, I'm back to this. While I get restarted, @damondouglas... Have you confirmed that these samples run? Have you run the formatter on them? Thanks!

kwalrath commented on dart_io_mini_samples/example/files_directories_and_symlinks/directories/creating_a_directory.dart:
```diff
@@ -8,11 +8,8 @@
-    .then((Directory directory) {
-      print(directory.path);
-  });
+  var directory = await new Directory('dir/subdir');
```
Shouldn't this be .create(recursive: true) so it'll create dir, as well as subdir?

kwalrath commented on dart_io_mini_samples/example/files_directories_and_symlinks/files/handling_errors_when_reading_a_file.dart:
```diff
@@ -12,14 +9,11 @@ void handleError(e) {
-      print('Could not read $filename');
-    });
+  try {
+    var file = await new File(filename).readAsString();
```
add print(file) to match original code?

kwalrath commented on dart_io_mini_samples/example/files_directories_and_symlinks/files/handling_errors_when_reading_a_file.dart:
```diff
@@ -12,14 +9,11 @@ void handleError(e) {
+  try {
+    var file = await new File(filename).readAsString();
+  } catch(e) {
+    handleError(e);    
```
Weird that it didn't use handleError before. Maybe delete handleError(e) here and have the old catchError code here, since it refers to filename?

kwalrath commented on dart_io_mini_samples/example/files_directories_and_symlinks/finding_the_type_of_a_filesystem_object.dart:
```diff
@@ -6,30 +6,31 @@
+    FileSystemEntityType type = await FileSystemEntity.type(entity.path);
+
+    String label;
+      switch (type) {
```
Is the switch statement indented too far?

kwalrath commented on dart_io_mini_samples/example/files_directories_and_symlinks/getting_the_parent_directory.dart:
```diff
@@ -7,12 +7,11 @@
-    });
+  Stream<FileSystemEntity> entityList = Directory.systemTemp.list(recursive: true, followLinks: false);
+
+  await for (FileSystemEntity entity in entityList) print(entity.parent.path);
```
I'm curious about whether the formatter would try to break this into multiple lines...

kwalrath commented on dart_io_mini_samples/example/files_directories_and_symlinks/symlinks/checking_if_a_path_represents_a_symlink.dart:
```diff
@@ -2,17 +2,17 @@
-    });
+  Stream<FileSystemEntity> entityList = systemTempDir.list(recursive: true, followLinks: false); 
+  await for (FileSystemEntity entity in entityList) {
+  	// Print the path only if it represents a symlink.
```
This looks like it's indented too far to me. (Maybe you used tabs instead of spaces?)

kwalrath commented on dart_io_mini_samples/example/http/adding_custom_headers.dart:
```diff
@@ -8,18 +8,17 @@
-              "${repo['forks_count']}");
-    }
-  });
+  var response = await http.get(url, headers : {'User-Agent':'Dart/1.0 (My Dart client)'});
```
`headers : {` should be `headers: {`, I think

kwalrath replied:
Not that that's your fault, since you didn't write that originally. :) dartfmt would catch stuff like this.

kwalrath commented on dart_io_mini_samples/example/http/handling_an_httprequest_error.dart:
```diff
@@ -2,12 +2,7 @@
-/// response is handled by the `then()` callback, and the error is handled by
-/// the `catchError()` callback.
-
+/// An HTTP request may return a response, or it may generate an error. 
```
I love how async-await makes this so much simpler!

kwalrath commented on dart_io_mini_samples/example/http/handling_an_httprequest_error.dart:
```diff
@@ -20,8 +15,12 @@ handleFailure(error) {
-    .catchError(handleFailure);
+main() async {
+  try {
+  	var response = await http.get("http://some_bogus_website.org");
```
looks over-indented to me.

kwalrath commented on dart_io_mini_samples/example/http/handling_an_httprequest_error.dart:
```diff
@@ -20,8 +15,12 @@ handleFailure(error) {
+  	var response = await http.get("http://some_bogus_website.org");
+  	handleSuccess(response);
+  }
+  catch(e) {
```
I expected `} catch(e) {` instead of having the `}` on its own line. But use whatever dartfmt produces.

kwalrath commented on dart_io_mini_samples/example/http/making_multiple_requests_to_the_same_server.dart:
```diff
@@ -18,16 +18,12 @@ printResponseBody(response) {
+  printResponseBody(response);
+  response = await client.get('${url}/doodles');
+  printResponseBody(response);
+  client.close();
```
to be exactly the same, shouldn't you put a try-catch around the code, and then put client.close() after that?

kwalrath commented on dart_io_mini_samples/example/http_server/implementing_a_Hello_World_http_server.dart:
```diff
@@ -2,20 +2,19 @@
+  var server = await HttpServer.bind(InternetAddress.LOOPBACK_IP_V4, 8080);
+  print("Serving at ${server.address}:${server.port}");
+  await for (HttpRequest request in server) {
+  	request.response
```
indentation looks off

kwalrath commented on dart_io_mini_samples/example/http_server/routing_requests_based_on_url_patterns.dart:
```diff
@@ -35,12 +35,11 @@ serveNotFound(req) {
+main() async {
+  var server = await HttpServer.bind(InternetAddress.LOOPBACK_IP_V4, 8080);
+  var router = new Router(server)
+  // Associate callbacks with URLs.
```
I think this line and the .. lines under it should be indented 2 more spaces.

kwalrath commented on dart_io_mini_samples/example/sockets/using_serversockets_server.dart:
```diff
@@ -8,12 +8,8 @@
+main() async {
+  var serverSocket = await ServerSocket.bind('127.0.0.1', 4041);
+  print('connected');
+  await for (var socket in serverSocket) socket.transform(UTF8.decoder).listen(print);
```
Did you look into getting rid of the second listen() here?

(It probably made the code longer, so keeping listen probably makes sense.)

kwalrath commented on dart_io_mini_samples/example/websockets/using_websockets_server.dart:
```diff
@@ -13,17 +13,15 @@ handleMsg(msg) {
+        var socket = await WebSocketTransformer.upgrade(req);
+        socket.listen(handleMsg);
+      }
+    }
```
Would it be possible/practical to make main() async and then use try-catch instead of onError below?

kwalrath commented:
I had a bunch of nitty little comments, but overall this looks great!

damondouglas commented:
Hello @kwalrath , thank you so much for your help and comments.  I will address those straight away.

damondouglas commented:
Hello @kwalrath , Thank you for your patience!  I scrubbed the code through `dartfmt`.  Pardon, I forgot this step.  I also changed some of the code according to your helpful comments.

kwalrath commented:
Sorry, I dropped the ball on this one. Let me take another look.

kwalrath commented on dart_io_mini_samples/example/http/making_multiple_requests_to_the_same_server.dart:
```diff
@@ -18,16 +18,16 @@ printResponseBody(response) {
+    printResponseBody(response);
+    client.close();
+  } catch (e) {
+    client.close();
```
You could use finally instead of duplicating client.close().

kwalrath commented on dart_io_mini_samples/example/websockets/using_websockets_server.dart:
```diff
@@ -6,24 +6,18 @@
-}
\ No newline at end of file
+main() async {
+  var server = await HttpServer.bind('127.0.0.1', 4040);
```
Should we have a try-catch around this to handle bind errors (or whatever other errors might occur)?

kwalrath commented:
A couple of minor comments, but this is already so much better I'd accept it as-is if you don't have time right now to make more changes.

damondouglas commented:
Thank you, again. Pardon, we are moving today to Seattle so I won't be able
to look for some time.
On Oct 1, 2015 5:34 PM, "Kathy Walrath" <notifications@github.com> wrote:

> A couple of minor comments, but this is already so much better I'd accept
> it as-is if you don't have time right now to make more changes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dart-lang/dart-samples/pull/93#issuecomment-144856660>
> .
>

kwalrath commented:
Good luck on the move! I'm going to accept this PR, and open a bug reminding us to consider the remaining comments.
